### DISCUSSION TITLE
Is there a way to automatically "clean" empty commits?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2481] [Q&A] [answered]
I seem to find myself with empty commits relatively frequently (probably largely because I'm still getting used to `jj` idioms).

Is there a way to globally `squash` all empty commits into their respective parents, or equivalently, `abandon` them (rebasing their children onto their parents)?

**Comment by martinvonz** [comment]
Try `jj log -r '(~empty())..'` to find empty commits that don't have non-empty descendants (or, to more directly match the revset, "commits that are not ancestors of non-empty commits"). If you want to git rid of those commits, just pass the same revset to `jj abandon`. (You may not want to squash them because that would rewrite their parents, which is probably not what you want if some of the parents exist on some remote.)

I'm also curious how you end up with the empty commits.

**Reply by BatmanAoD** [reply]
I'll try to make a mental note to check `op log` next time I see it, but I suspect it's usually from using `checkout` multiple times in a row without making intervening changes.

**Reply by ilyagr** [reply]
I run into this issue (excessive empty commits) as well. I have never gotten to debugging it properly or found a reproducible way to cause it, would appreciate any discoveries you make.

**Reply by elasticdog** [reply]
IIRC, when I've gotten empty commits that I had to abandon in the past, it was mostly from:

1. working on a branch
2. pushing that branch and opening a PR
3. rebase merging that PR
4. then running `jj git fetch` without remembering to run `jj co main` first (so I was still on the feature branch locally)

Once I got into the habit of always running `jj co main` after pushing my branch, those rebased commits have been cleaned up automatically during the fetch process.

**Reply by BatmanAoD** [reply]
`fetch` creates empty commits depending on whether or not you're on the trunk branch?? That's... baffling, frankly. That's _definitely_ one of the reasons I have empty commits showing up.

`git fetch` only updates remotes, and does _nothing_ with the working copy. Why is `jj` doing more than fetching commits?

**Reply by BatmanAoD** [reply]
...but `checkout` also creates a new commit, and I don't know a simpler way to switch branches/commits, so I think that's also one of the reasons I'm getting extra commits.

(Part of "simpler" here includes "easy to remember.")

**Reply by BatmanAoD** [reply]
Please forgive a bit of soapboxing.

If I were designing a system like this, I would follow these guidelines:

* There's no such thing as a fully `empty` commit, with no description _and_ no changes from its parent.
  * This implies that `@` would always be the latest "real" commit, never an empty commit.
  * Edit: it's fine that the root commit is empty.
* `checkout` never introduces new commits, and always changes the state of the working copy.
* `fetch` never changes the state of the working copy, but does introduce new commits if it finds any.

**Reply by martinvonz** [reply]
> There's no such thing as a fully `empty` commit, with no description _and_ no changes from its parent.
> * This implies that `@` would always be the latest "real" commit, never an empty commit.

This comes up once in a while. It simplifies the model quite a bit if the working copy is always a real commit. Otherwise we probably have to pretend that `@` is some virtual commit until it becomes real. Take this example:

```
$ jj checkout main
$ jj diff -r @
$ jj describe -r @ -m foo
$ jj diff -r @
$ echo foo >> README.md
$ jj diff -r @
```

Unless `@` is a virtual commit or a real commit there, then some of those (maybe just the first one) would instead show the diff in commit pointed to by `main`. That behavior would probably be surprising, so I think we would need a real commit or a virtual commit. I haven't thought much about it, but it seems like having a real commit would be simpler. For example, if it's a virtual commit, do you give it a change id? Is that change id preserved when you do `jj describe -r @`?

> * `checkout` never introduces new commits, and always changes the state of the working copy.

Unless the content in the two commits is the same, I assume. This seems like basically the same as the bullet above.

> * `fetch` never changes the state of the working copy, but does introduce new commits if it finds any.

The reason `jj git fetch` can change the state of the working copy is that it tries to be helpful and propagate deleted commits from the remote. Let's say you have this:

```
@ xxz
| some local work
o xxy feature
|
o xxx main
```

If the feature branch was now deleted from the remote, then `jj git fetch` we interpret that as saying that the commits that became unreachable on the server should also be deleted locally. So we abandon `xxy` (and delete branch `feature`), which means your working-copy commit (`xxz`) gets updated.


I hope that at least clarifies the reasoning behind the current behavior.

**Reply by ilyagr** [reply]
> I don't know a simpler way to switch branches/commits, so I think that's also one of the reasons I'm getting extra commits.

You can get close to the workflow you suggest by using `jj edit` instead of `jj checkout`. That approach has pluses and minuses, but is certainly worth trying out to see if you prefer it, and worth being aware of even if you don't use it all the time (I, personally, use it a small minority of the time, mainly because it makes some conflicts from race conditions more likely, when an edit to the working copy can cause a sequence of rebases. With jj, these should never lose data, but can be a bit annoying).

**Update:** `jj edit` won't work on immutable commits, so e.g. `jj edit main` shouldn't work. So, you'll need to use `jj new` or `jj checkout` occasionally. You can `jj duplicate main` and then `jj edit` the result, if you like, though this is only useful in special cases.

**Reply by BatmanAoD** [reply]
I have to admit, it was not immediately obvious to me what the "expected" behavior of that sequence of `jj diff` commands would even be. This is largely because it's already surprising to me that `diff -r c` is equivalent to `diff --from c- --to c`, so much so that I haven't actually used `diff` that way yet and have been exclusively using `--from` and `--to` so that I'd be certain I know what I'm looking at. Having the implicit side of the comparison be "the working copy" seems obvious and intuitive (...though admittedly that could be from exposure to git); having it be _the parent commit_ seems...unexpected. So, in my hypothetical model where `@` is the current non-empty commit, I agree that `jj`'s `diff -r @` wouldn't really make any sense. I'm just...not really sure what sense it makes in any other model, either. 😅 

That said, I do fairly sometimes use `git log -p -1 <commit>`, which is arguably an even less intuitive way to get similar behavior, though at least in that case the use of `log` implies automatic traversal of parent commits (...unlike in `jj`).

So... what would I expect `jj checkout main; jj log -p -r @-::@` to display in that mental model? I guess I'd expect it to show the same diff as `git checkout main; git log -p -1`, i.e. `jj diff -r main`.

Coming back to `diff -r`... I guess it's "the changeset applied to the previous commit to create this commit." (I.e. exactly the content constituting the "commit" in a patch-based VCS like SVN.) But even with that mental model, I'm not sure why the following output would be surprising:

```
$ jj checkout main
$ jj diff -r @
<same as jj diff -r main>
$ jj describe -r @ -m foo
$ jj diff -r @
<empty>
$ echo foo >> README.md
$ jj diff -r @
<added foo>
```

In general, empty output from `diff` seems surprising in this model unless you've explicitly created an empty commit.

----

It occurs to me that "Working copy changes" section of the `status` output in `jj` is effectively `jj diff -s -r @`, which is kind of interesting. That obviously wouldn't be true in my hypothetical model, but honestly in a model like that, it's not totally clear what "working copy changes" would even mean. The only useful definition I can think of off-hand is something like "the set of all changes between the working copy and the most recent branch-tips," but that's...pretty complicated, as definitions go. Honestly, if I were actually designing a system like this, I would probably define some first-class command, separate from both `diff` and `log`, that behaves somewhat like `git log -p`/`jj diff -r` but represents more generally a diff over a nonempty sequence of commits, and then make `status` an alias for some particular invocation of the `changes` command.

----

...in any case: I'm not really trying to argue that `jj` should actually adopt a different model for what `@` means and whether empty commits can exist. I think my hypothetical model is too different from how `jj` exists. I'm just trying to give some insight into the things that are making `jj`'s model kind of hard to grok, as a longtime `git` user (with all the...er...maladaptation that implies).

**Reply by BatmanAoD** [reply]
Thanks for explaining that behavior of `fetch`. I have to admit, I still find it...pretty unintuitive. Does that mean that in your example, `jj fetch` will essentially apply `git revert xxy`, deleting those changes? Is there a way to suppress that behavior?

**Reply by martinvonz** [reply]
> Does that mean that in your example, `jj fetch` will essentially apply `git revert xxy`, deleting those changes?

Yes, that's correct. Technically, what happens is that the commits get abandoned, which results in the working-copy commit (and any other commits on top) getting rebased off of the abandoned commits. Then the working copy (the files you edit in your editor) gets updated to the new working-copy commit.

> Is there a way to suppress that behavior?

Not really in any native way, but if you're in a colocated repo, you can use regular `git fetch`. I think that will not even delete the remote-tracking branch unless you pass `--prune`. If you do pass that, then the commits will still remain once jj imports the new state.

**Reply by BatmanAoD** [reply]
I have to admit, even having given myself time to sleep on that, it seems absolutely wild to me. I don't think I would _ever_ want that behavior, certainly not automatically, and _definitely_ not as part of the only way to download data from the remote.

**Reply by BatmanAoD** [reply]
I guess I can see a case for that being the right behavior for `pull`.

**Reply by BatmanAoD** [reply]
If there are multiple branches that point to `xxy`, and only one of them is deleted remotely, the commit won't be abandoned, right?

**Reply by martinvonz** [reply]
> If there are multiple branches that point to `xxy`, and only one of them is deleted remotely, the commit won't be abandoned, right?

Correct.

> I have to admit, even having given myself time to sleep on that, it seems absolutely wild to me. I don't think I would _ever_ want that behavior, certainly not automatically, and _definitely_ not as part of the only way to download data from the remote.

I'm open to making it configurable. I don't think I've heard complaints about it before, but it's certainly possible that others would also want it. Mind filing a feature request?

**Reply by BatmanAoD** [reply]
>  it's certainly possible that others would also want it.

Well, I told a friend and he said "y'know what, just gonna delete the `jj` binary while I'm reading this" 😆 

Feature request: https://github.com/martinvonz/jj/issues/2504

**Reply by martinvonz** [reply]
> Well, I told a friend and he said "y'know what, just gonna delete the `jj` binary while I'm reading this" 😆

There's obviously selection bias in the complaints we hear. Maybe most people simply uninstall jj and we don't hear from them again. So thanks for letting us know.

What do you typically do with the (remote-)deleted branches in Git? I'm guessing you would still delete them eventually, and you're just saying that you prefer to do it manually/explicitly?

**Reply by BatmanAoD** [reply]
I'm happy to see _branches_ deleted, I'm just uncomfortable with commits still reachable from the working copy being deleted. I use `git fetch --prune` far more often than `git fetch`, and I wish that it would delete local tracking branches as well (unless there are commits on the local branch that weren't on the last-seen version of the remote branch, I guess).

I'm possibly still confused about the `jj fetch` behavior. Is it only the state of the working copy that can cause a rebase, or can any arbitrary commit-sequence potentially be rebased if a branch points somewhere into its ancestor tree and is then deleted? That can't be right, can it?

**Reply by BatmanAoD** [reply]
...I guess a more general rule is: a branch is just a tag, so I don't ever expect deleting a branch to cause commits to be deleted, much less for new commits (the rebase sequence) to be created automatically. I would expect it to have exactly the same behavior as if I deleted a tag.

**Reply by martinvonz** [reply]
There are many things that fed into the current behavior:

* I copied many ideas from hg. One of them is anonymous branches/heads.
* I initially copied "changeset evolution" from hg. I later realized that first-class conflicts means that we can automatically rebase commits right away instead of having something like hg's explicit `hg evolve` command.
* When a commit has been obsoleted (e.g. `hg prune`),`hg evolve` will rebase descendants onto the parent. Similarly, jj's automatic rebasing will rebase descendants onto the parent when you `jj abandon` a commit
* If the commit was made obsolete on a remote, then `hg pull` (the `git fetch`-like thing) would propagate that locally, and a subsequent `hg evolve` would fill in the same was as if it had been obsoleted locally to start with. There's no native jj remote yet, but if there were, and a commit had been abandoned on the remote, I would like that to propagate to the local repo like it does with hg
* Git remotes don't have a concept of obsolete or hidden commits. Instead, they define anything that's reachable from a ref as visible. `jj git fetch` takes that definition and uses it for figuring out which commits have been "abandoned" (i.e. become unreachable) on the remote.

I hope that helps clarify the current behavior. I don't mean to say that we can't change it.

**Reply by necauqua** [reply]
As much as I love jj, this behavior is indeed strange - I just never fetched a branch deletion where locally I have child commits of that branch

Like, deleting a tip is fine, but abandoning (and thus rebasing children) the middle commit just doesn't make that much sense to me at all, it's just weird, almost like a bug, or a git brain thing

And yeah, currently in jj branches are literally lightweight tags that are only needed for git remote interop - me being fine with deletion of the tip is a git brain thing - and whoops, I just realised that just the tip also doesn't make much sense, I was thinking more about like a branch in git (ancestors upto the trunk?. not even well-defined in my head)

tl;dr I never realised this was a thing and I completely agree with the batman man

**Reply by BatmanAoD** [reply]
So, having finally run into this personally and found it to be exactly as bad as I feared (...I squash my PRs when I merge them, which exacerbated things; and this behavior isn't just on `jj git fetch` but on _every command_ that imports from `.git`, including `jj status`), I think the big question that still remains in my mind is: if deleting a commit requires rebasing child commits, then by definition, _that commit was reachable_, so...why is it being deleted?

**Comment by ilyagr** [comment, marked as answer]
I hadn't even realized this was possible, but it [apparently is](https://martinvonz.github.io/jj/prerelease/revsets/).

Try `jj log -r 'empty() ~ merges()'`. If you like the result, you can `jj abandon 'empty() ~ merges()'`. If you include merges, there is a minor bug with `abandon` (https://github.com/martinvonz/jj/issues/2482) that will mess up the topology of the repo a bit.

At the moment, ~I don't see a way to exclude commits with non-empty descriptions.~

**Update:** Actually, @yuja implemented that as well.

`jj log -r '(empty() ~ merges()) & description(exact:"")'`

**Reply by martinvonz** [reply]
Good point about non-empty descriptions. You can use `description(exact:"")` to match commits with empty descriptions.

**Reply by BatmanAoD** [reply]
This also needs `~ root()`.

**Reply by ezzieyguywuf** [reply]
> This also needs `~ root()`

For my future lazy self here's the full incantation:

```
jj log -r '(empty() ~ merges()) & description(exact:"") ~ root()'
```
